### DISCUSSION TITLE
add `MutDeque.insert`

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#8476] [Ideas]
There is no operation to insert into a `MutDeque` with an arbitrary index.

**Comment by JonathanStarup** [comment]
"you should just use MutList" is not an answer since the loses constant time pop from the front

**Comment by magnus-madsen** [comment]
But queues tend not to support such an operation.

**Comment by JonathanStarup** [comment]
`MutDeque` is not an interface, though - this implementation can easily add a linear time insert operation

`insert` also exists in the theoretical queue interface, doing a sequence of pops, then push the element, then push the old elements. This is just a shorthand for that. you have worst case n/2 time if you check which end is closer, compared to mutlist that has worst case n

**Comment by mlutze** [comment]
I think this brings up the question of interfaces vs implementations. Deque is an interface. Maybe this should be called ArrayDeque or something that indicates its implementation.

**Comment by mlutze** [comment]
Then we add a queue trait that it implements

**Comment by JonathanStarup** [comment]
I want to add that https://github.com/flix/flix/pull/10141 has an application that would benefit from ~~this~~ `MutDeque.nth`

**Reply by magnus-madsen** [reply]
I don't think they can afford to use MutList; two much indirection for what is the core data structure.

**Reply by JonathanStarup** [reply]
This is not MutList, its MutDeque. And actually they want random lookup, not insertion, but same point
